embedded-graphics/src/pixelcolor/raw/mod.rs

```diff
@@ -0,0 +1,225 @@
+    fn into_inner(self) -> Self::Base;
+
+    /// Converts a `u32` into a `RawData` type.
+    ///
+    /// This method can be used to generically construct all `RawData` types from
+    /// the same integer type. If the width of the `RawData` type is less than
+    /// 32 bits only the least significant bits are used.
+    fn from_u32(value: u32) -> Self;
+}
+
+/// Dummy implementation for `()`.
+///
+/// `()` can be used as [`PixelColor::Raw`] if raw data conversion isn't required.
+///
+/// [`PixelColor::Raw`]: ../trait.PixelColor.html#associatedtype.Raw
+impl RawData for () {
+    type Base = ();
+
+    const BITS_PER_PIXEL: usize = 0;
+
+    fn into_inner(self) {}
+
+    fn from_u32(_value: u32) {}
+}
+
+impl private::Sealed for () {}
+
+macro_rules! impl_raw_data {
+    ($type:ident : $base_type:ident, $bpp:expr, $mask:expr, $bpp_str:expr) => {
+        #[doc = $bpp_str]
+        #[doc = "per pixel raw data."]
+        #[doc = ""]
+        #[doc = "See the [module-level documentation] for more information."]
+        #[doc = ""]
+        #[doc = "[module-level documentation]: index.html"]
+        #[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+        pub struct $type($base_type);
+
+        impl $type {
+            /// Creates a new color from the least significant
+            #[doc = $bpp_str]
+            /// of value.
+            pub const fn new(value: $base_type) -> Self {
+                $type(value & $mask)
+            }
+        }
+
+        impl RawData for $type {
+            type Base = $base_type;
+
+            const BITS_PER_PIXEL: usize = $bpp;
+
+            fn into_inner(self) -> Self::Base {
+                self.0
+            }
+
+            fn from_u32(value: u32) -> Self {
+                #[allow(trivial_numeric_casts)]
+                Self::new(value as $base_type)
+            }
+        }
+
+        impl From<$base_type> for $type {
+            fn from(value: $base_type) -> Self {
+                Self::new(value)
+            }
+        }
+
+        impl private::Sealed for $type {}
+    };
+}
+
+impl_raw_data!(RawU1: u8, 1, 0x01, "1 bit");
+impl_raw_data!(RawU2: u8, 2, 0x03, "2 bits");
+impl_raw_data!(RawU4: u8, 4, 0x0F, "4 bits");
+impl_raw_data!(RawU8: u8, 8, 0xFF, "8 bits");
+impl_raw_data!(RawU16: u16, 16, 0xFFFF, "16 bits");
+impl_raw_data!(RawU24: u32, 24, 0xFFFFFF, "24 bits");
+impl_raw_data!(RawU32: u32, 32, 0xFFFFFFFF, "32 bits");
+
+/// Raw data byte order.
+pub trait ByteOrder: private::Sealed {}
+
+/// Little endian byte order marker.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum LittleEndian {}
+
+impl ByteOrder for LittleEndian {}
+impl private::Sealed for LittleEndian {}
+
+/// Big endian byte order marker.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum BigEndian {}
+
+impl ByteOrder for BigEndian {}
+impl private::Sealed for BigEndian {}
+
+mod private {
+    /// Sealed trait to prevent implementation of traits in other crates.
+    pub trait Sealed {}
```

Comment:
Traits like `RawData` can't effectively be implemented outside the crate because they are tightly integrated with the other code. The `Sealed` trait prevents users from trying and allows us to change the internal API without having to worry about breaking code in other crates. The `byteorder` crate uses the same trick.

Reply:
Right, that makes sense. I assume we can't just keep it private because it's exported in a public interface?

Reply:
Exactly.

A better explanation why this is a good idea is: https://rust-lang-nursery.github.io/api-guidelines/future-proofing.html#sealed-traits-protect-against-downstream-implementations-c-sealed